### PR TITLE
[dynamo] Support TypedDict

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138378
* __->__ #138359

Seen in vLLM.

Fixes https://github.com/pytorch/pytorch/issues/132629
Fixes https://github.com/pytorch/pytorch/issues/133613

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -369,6 +369,14 @@ def call_function(
                 args[0],
                 mutable_local=MutableLocal(),
             )
+        elif is_typeddict(self.value):
+            if self.value.__optional_keys__:
+                unimplemented("typeddict with optional keys not supported")
+            if len(args):
+                unimplemented("TypedDict only supports kwargs")
+            keys_vt = [variables.ConstantVariable.create(key) for key in kwargs]
+            dict_items = dict(zip(keys_vt, kwargs.values()))
```
Comment (Skylion007) [inline torch/_dynamo/variables/user_defined.py:378]:
```suggestion
            dict_items = {variables.ConstantVariable.create(key):v for key, v in kwargs.items()}
```

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -14,6 +14,16 @@
...
 from typing import Dict, Generic, List, TYPE_CHECKING
 
+
+try:
+    # Not present in all python versions
+    from typing import is_typeddict
+except ImportError:
+
+    def is_typeddict(t):
+        return False
```
Comment (Skylion007) [inline torch/_dynamo/variables/user_defined.py:24]:
```suggestion
      from typing_extensions import typing-extensions
```

File: torch/_dynamo/variables/user_defined.py [hunk 3]
```diff
@@ -369,6 +370,16 @@ def call_function(
                 args[0],
                 mutable_local=MutableLocal(),
             )
+        elif is_typeddict(self.value):
+            if self.value.__optional_keys__:
+                unimplemented("typeddict with optional keys not supported")
```
Comment (Skylion007) [inline torch/_dynamo/variables/user_defined.py:375]: Be consistent about Capitalization of TypedDict?

**Review by Skylion007 — DISMISSED** [review]

File: test/dynamo/test_repros.py [hunk 4]
```diff
@@ -6107,6 +6107,42 @@ def inject_parameters(module, cls):
...
 
+    def test_typed_dict(self):
+        class LlavaImagePixelInputs(TypedDict):
+            type: Literal["pixel_values"]
+            data: torch.Tensor
+            """Shape: `(batch_size, num_channels, height, width)`"""
+
+        def fn(x, y):
+            obj = LlavaImagePixelInputs(type=int, data=y)
+            return x * obj["data"]
```
Comment (Skylion007) [inline test/dynamo/test_repros.py:6118]: Shouldn't we add some test for mutations of the keys?

File: torch/_dynamo/variables/user_defined.py [hunk 5]
```diff
@@ -369,6 +370,16 @@ def call_function(
...
+        elif is_typeddict(self.value):
+            if self.value.__optional_keys__:
+                unimplemented("TypedDict with optional keys not supported")
+            if len(args):
+                unimplemented("TypedDict only supports kwargs")
+            dict_items = {
+                variables.ConstantVariable.create(key): value
+                for key, value in kwargs.items()
+            }
+            return variables.ConstDictVariable(dict_items, mutable_local=MutableLocal())
```
Comment (XuehaiPan) [inline torch/_dynamo/variables/user_defined.py:382]: The `__new__` method for a TypedDict class is `dict.__new__`. It supports `TypedDictClass(mapping, **kwargs)` and `TypedDictClass(seq2, **kwargs)`.

```python
In [1]: from typing import TypedDict

In [2]: class MyDict(TypedDict):
   ...:     foo: int
   ...:     bar: float
   ...:     

In [3]: MyDict([('foo', 1), ('bar', 0.0)])
Out[3]: {'foo': 1, 'bar': 0.0}

In [4]: MyDict({'foo': 1, 'bar': 0.0})
Out[4]: {'foo': 1, 'bar': 0.0}

In [5]: MyDict([('foo', 1)], bar=10.0)
Out[5]: {'foo': 1, 'bar': 10.0}
```

Just use `dict.__new__` here:

```suggestion
            return variables.BuiltinVariable(dict).call_dict(tx, args, kwargs)
```

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**